extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/dev/HibernateOrmDevJsonRpcService.java

```diff
@@ -18,4 +46,100 @@ public int getNumberOfNamedQueries() {
         return getInfo().getNumberOfNamedQueries();
     }
 
+    private Optional<HibernateOrmDevInfo.PersistenceUnit> findPersistenceUnit(String persistenceUnitName) {
+        return getInfo().getPersistenceUnits().stream().filter(pu -> pu.getName().equals(persistenceUnitName)).findFirst();
+    }
+
+    public DataSet executeHQL(String persistenceUnit, String hql, Integer pageNumber, Integer pageSize) {
+        if (!isDev) {
+            return errorDataSet("This method is only allowed in dev mode");
+        }
+
+        if (!hqlIsValid(hql)) {
+            return errorDataSet("The provided HQL was not valid");
+        }
+
+        Optional<HibernateOrmDevInfo.PersistenceUnit> pu = findPersistenceUnit(persistenceUnit);
+        if (pu.isEmpty()) {
+            return errorDataSet("No such persistence unit: " + persistenceUnit);
+        }
+
+        //noinspection resource
+        SessionFactoryImplementor sf = pu.get().sessionFactory();
+
+        // Check the connection for this persistence unit points to an allowed datasource
+        ConnectionProvider connectionProvider = sf.getServiceRegistry().requireService(ConnectionProvider.class);
+        if (connectionProvider instanceof QuarkusConnectionProvider quarkusConnectionProvider) {
+            if (!isAllowedDatabase(quarkusConnectionProvider.getDataSource())) {
+                return errorDataSet("The persistence unit datasource points to a non-allowed datasource " +
+                        "(by default, only local databases are allowed).");
```

Comment:
Maybe make this actionable by mentioning the configuration property to allow non-local DBs?

Reply:
Changed to a more verbose message, mentioning the configuration property.